cl: Fix properties syncer test

michael-redpanda opened the pull request:
Update test to wait for the expected topic properties to be set on the shadow topic before starting the test.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [X] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

IoannisRP commented:
i am not sure i understand what the difference is here.

Wouldn't the `chunked_hash_map::operator==` do exactly the same checks that you are doing manually?

In what case would the new code produce different results?

michael-redpanda committed:
cl/tests: Fix flaky properties syncer test

Await for all expected topic properties to be present in the mirror
topic before starting the test.

Signed-off-by: Michael Boquard <michael@redpanda.com>

IoannisRP commented on src/v/cluster_link/tests/topic_properties_syncer_test.cc:
```diff
@@ -74,6 +82,22 @@ class topic_properties_syncer_test : public seastar_test {
         co_await fixture()->upsert_link(get_default_metadata());
         fixture()->get_cluster_mock().add_topic(
           test_topic, 1, 1, kafka::topic_authorized_operations(0x508));
+
+        // Wait for all expected topic properties to be present in the mirror
+        // topic
+        RPTEST_REQUIRE_EVENTUALLY_CORO(5s, [this] {
+            auto link = fixture()->find_link_by_name(test_link_name);
+            const auto& mirror_topics = link->get().state.mirror_topics;
+            auto mirror_topic_it = mirror_topics.find(test_topic);
+            if (mirror_topic_it == mirror_topics.end()) {
+                return false;
+            }
+            return std::ranges::all_of(
+              expected_properties_to_be_present, [&](const auto& property) {
+                  return mirror_topic_it->second.topic_configs.contains(
+                    property);
+              });
+        });
```
Am I reading this correctly?

Previously, we didn't wait enough time in the fixture set up, so when the time came to compare, there was a missing entry (`"redpanda.remote.allowgaps"`) in the `properties.topic_configs`.
Now with the extra waiting time that ensures all relevant values are present, it will either wait till they are there (expected path), or the timeout will fail.

michael-redpanda replied:
yes

bharathv reviewed (approved):
update cover letter?

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--261183f8-43e2-4832-9304-994301a21458-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73634">build#73634</a></summary>


| test_class                   | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    | test_history                                                                                                                                                                  |
|:-----------------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerGroupOffsetResetTest | test_stress_consumer_group_commits   | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73634#0199ba03-beac-4733-b91b-965778562a6b | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerGroupOffsetResetTest&test_method=test_stress_consumer_group_commits |
| DataMigrationsApiTest        | test_creating_and_listing_migrations | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73634#0199ba03-bea6-4948-8afe-bf8a1206ec28 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations      |
</details>
<!--/body-->
